Mathlib/Algebra/Module/LinearMap.lean

```diff
@@ -1023,6 +1052,51 @@ instance module : Module S (M →ₛₗ[σ₁₂] M₂) where
 instance [NoZeroSMulDivisors S M₂] : NoZeroSMulDivisors S (M →ₛₗ[σ₁₂] M₂) :=
   coe_injective.noZeroSMulDivisors _ rfl coe_smul
 
+variable {S : Type*} [Semiring S] [Module S M]
+
+instance : Module Sᵈᵐᵃ (M →+ M₂) where
+  add_smul s s' f := AddMonoidHom.ext fun m ↦ by
+    simp_rw [AddMonoidHom.add_apply, DomMulAct.smul_addMonoidHom_apply, ← map_add, ← add_smul]; rfl
+  zero_smul _ := AddMonoidHom.ext fun _ ↦ by
+    erw [DomMulAct.smul_addMonoidHom_apply, zero_smul, map_zero]; rfl
+
+instance [SMulCommClass R S M] : Module Sᵈᵐᵃ (M →ₛₗ[σ₁₂] M₂) where
+  add_smul _ _ _ := ext fun _ ↦ by simp_rw [add_apply, smul_apply', ← map_add, ← add_smul]; rfl
+  zero_smul _ := ext fun _ ↦ by erw [smul_apply', zero_smul, map_zero]; rfl
+
+abbrev _root_.AddMonoidHom.domModule' : Module Sᵐᵒᵖ (M →+ M₂) := inferInstanceAs (Module Sᵈᵐᵃ _)
+
+abbrev domModule' [SMulCommClass S R M] : Module Sᵐᵒᵖ (M →ₛₗ[σ₁₂] M₂) :=
+  have := SMulCommClass.symm S R M; inferInstanceAs (Module Sᵈᵐᵃ _)
+
+attribute [local instance] AddMonoidHom.domModule' domModule'
+
+@[simp] lemma _root_.AddMonoidHom.domModule_smul_apply' (s : Sᵐᵒᵖ) (f : M →+ M₂) (m : M) :
+    (s • f) m = f (s.unop • m) := rfl
+
+@[simp] lemma domModule_smul_apply' [SMulCommClass S R M] (s : Sᵐᵒᵖ) (f : M →ₛₗ[σ₁₂] M₂) (m : M) :
+    (s • f) m = f (s.unop • m) := rfl
+
+variable {S : Type*} [Semiring S] [Module Sᵐᵒᵖ M]
+
+/-- If `M` is a right `S`-module and `M₂` is an commutative monoid, then `M →+ M₂` is
+  a left `S`-module. This is not an instance because `S` may also act on `M₂`. -/
+abbrev _root_.AddMonoidHom.domModule : Module S (M →+ M₂) :=
```

Comment:
Is there a reason we need this copy of the instance without the synonym?

Reply:
This one actually cancels `dma` with `mop` to obtain a `S`-action from a `S^mop`-action. The reason is that I want to write `N →ₗ[R] M →+ P` for `N` a `R`-module and `M` a `R^mop`-module, and I don't think you'd like to give `N` a `R^mop^dma`-module instance ... Maybe we could use semilinear maps, but in this file simply register `domModule` as local instances works nicely.

`domModule'` is the version that simply changes `dma` to `mop`, which allows us to write `M →ₗ[R^mop] N →+ P` in the same setting.

Reply:
Semilinear maps are indeed what I was thinking of here.